src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs

```diff
@@ -395,7 +396,7 @@ public sealed class Options : ArgumentsCore
         internal static class DefaultArguments
         {
             internal const bool AllowQuoting = true;
```

Comment:
>AllowQuoting [](start = 32, length = 12)

I feel like this needs to be changed too.

Reply:
Absolutely. Please take a look at Iteration 3.

---
In reply to: [258240404](https://github.com/dotnet/machinelearning/pull/2630#discussion_r258240404) [](ancestors = 258240404)

Reply:
I feel like having quoting on by default is better. Mainly this helps with CSV files w/ text columns (commas in the quoted strings). 

 | row | needs quotes |
|--|--|
| `true`, `"Schrödinger's cat walks into a bar, and doesn't."` |  yes | 
| `false`, `Wanted: Schrodinger's cat. Dead and alive.` | no |

I see more datasets with gains from default **on** quoting, then hurt by it.

Reply:
I'd like to have non-standard settings to be non-default. There is no standard on parsing quoted strings.

---
In reply to: [258341863](https://github.com/dotnet/machinelearning/pull/2630#discussion_r258341863) [](ancestors = 258341863)

Reply:
To be clear, my recommendation is to keep  `AllowQuoting` defaulting to `true`. 

General reason is that it more datasets will work. The impact of the parameter falsely on, is less than falsely off. 

/cc @TomFinley: thoughts?

Reply:
I'm actually rather surprised to hear you say this @justinormont, since if I think back, at least internally I recall you being one of the most frequent "victims" of the quoting. Indeed it was one of the most frequent questions we got over the years.

I think what you said would be true for a TSV and CSV parser, to adhere to that quasi-format. However this isn't a TSV or CSV parser, and differs from it in some important respects. I'd agree that an `IDataReader` implementation should adhere to the format of that, but that argument doesn't really apply here.

I'm not motivated exactly by any desire to read CSV format (this thing doesn't do that anyway, so, you can continue quoting as many APIs to read CSV format as you like), but more motivated by what I've perceived as more accidental problems caused by having this default *on* than would have been solved had we left it *off*.

Reply:
This is incorrect. Exactly zero of the issues I ran into would be solved by turning off quoting. 

The issue was our quoting support did not go far enough. I main issue I faced was that we haven't added support for newlines in the quoted string. I believe the argument at the time was that we treat lines independently for speed. 

Most datasets are in CSV/TSV format. We should be supporting the datasets of our users.

Reply:
If quote is not fully-functioning, it'd be better to turn it off by default and have another TSC & CSV reader implementation, I guess.

---
In reply to: [258668384](https://github.com/dotnet/machinelearning/pull/2630#discussion_r258668384) [](ancestors = 258668384)

Reply:
> This is incorrect. Exactly zero of the issues I ran into would be solved by turning off quoting.

I don't think that's true. I recall on multiple occasions seeing you run experiments where you were getting these warnings and you just sort of ignored it, even though it was probably corrupting your dataset. It definitely wasn't beneficial on the whole. It isn't a matter of it "not go far enough," because it wasn't intended to do that at any point. I believe this has also been explained to you multiple times by multiple people. At any rate, it again reinforces the point that this is not a CSV reader, which I believe I also mentioned in my reply.

There's also the issue that there are other people in the world, and on the whole if I look through my DRI email list, I see way more instances of turning quoting off than I see instances where someone somehow accidentally stumbled into using our quoting system.

So, off.

Reply:
I see this as lesser of two evils. 

Default ON:
* Has the potential to cause read errors (dropped rows) when there are newlines or quotes in the quoted string _(generally doesn't matter)_
* Non-text datasets are unaffected

Default OFF:
* Most text datasets fail _(bad)_
* Non-text datasets are unaffected

Let me know if I'm wrong, but I think the argument more directly is:
The user should be forced to mung their dataset in to the form we are expecting (eg: no commas/newlines/escaped-quotes in a string) vs. we should do our best on the data the user hands us (expectation of dirty data).

Reply:
To me it's fairly simpler than that. If I think of the user questions and scenarios that I've handled with text loader over the years, and I tally up (1) those datasets that conformed to *our* scheme of quoting and (2) those that did not, there's a really clear tendency. I also think of all the problems it's caused over the years.